RevenueCatUI/Plugins/iOS/RevenueCatUI.m

```diff
@@ -1,31 +1,170 @@
-    NSLog(@"[RevenueCatUI][iOS] presentPaywallIfNeeded(entitlement=%@, offering=%@, closeButton=%@)",
-          requiredEntitlementIdentifier ? [NSString stringWithUTF8String:requiredEntitlementIdentifier] : @"<null>",
-          offeringIdentifier ? [NSString stringWithUTF8String:offeringIdentifier] : @"<null>",
-          displayCloseButton ? @"true" : @"false");
-    if (callback) {
-        callback("NOT_PRESENTED|Stub: no native UI");
+static NSString *RCUINormalizedResultToken(NSString *resultName) {
+    if (resultName.length == 0) {
+        return @"ERROR";
     }
+
+    NSString *normalized = [[[resultName stringByReplacingOccurrencesOfString:@"_" withString:@""]
+                                       stringByReplacingOccurrencesOfString:@"-" withString:@""]
+                                      lowercaseString];
+
+    if ([normalized isEqualToString:@"purchased"]) {
+        return @"PURCHASED";
+    }
+    if ([normalized isEqualToString:@"restored"]) {
+        return @"RESTORED";
+    }
+    if ([normalized isEqualToString:@"cancelled"] || [normalized isEqualToString:@"usercancelled"]) {
+        return @"CANCELLED";
+    }
+    if ([normalized isEqualToString:@"notpresented"] || [normalized isEqualToString:@"notshown"]) {
+        return @"NOT_PRESENTED";
+    }
+    if ([normalized isEqualToString:@"error"] || [normalized isEqualToString:@"failed"]) {
+        return @"ERROR";
+    }
+
+    return [resultName uppercaseString];
+}
+
+static BOOL RCUIRuntimeSupportsPaywalls(void) {
+    if (@available(iOS 15.0, *)) {
+        return YES;
+    }
+    return NO;
+}
+
+static void RCUIInvokeCallback(RCUIPaywallResultCallback callback, NSString *token, NSString *message) {
+    if (callback == NULL) {
+        return;
+    }
+
+    NSString *payload = token ?: @"ERROR";
+    if (message.length > 0) {
+        payload = [payload stringByAppendingFormat:@"|%@", message];
+    }
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+        callback([payload UTF8String]);
+    });
+}
+
+static BOOL RCUIEnsureReady(RCUIPaywallResultCallback callback) {
+    if (!RCUIRuntimeSupportsPaywalls()) {
+        RCUIInvokeCallback(callback, @"NOT_PRESENTED", @"Requires iOS 15.0+");
+        return NO;
+    }
+
+    if (!RCPurchases.isConfigured) {
+        RCUIInvokeCallback(callback, @"ERROR", @"PurchasesNotConfigured");
+        return NO;
+    }
+
+    return YES;
 }
 
-// No Customer Center in this stub
+static NSMutableDictionary *RCUICreateOptionsDictionary(NSString *offeringIdentifier, BOOL displayCloseButton) {
+    NSMutableDictionary *options = [NSMutableDictionary new];
+    options[kRCUIOptionDisplayCloseButton] = @(displayCloseButton);
+
+    if (offeringIdentifier.length > 0) {
+        options[kRCUIOptionOfferingIdentifier] = offeringIdentifier;
+        options[kRCUIOptionPresentedOfferingContext] = @{ kRCUIPresentedOfferingIdentifierKey: offeringIdentifier };
```

Comment:
`PresentedOfferingContext` is a bit more complex than this, since we would need to pass whatever `PresentedOfferingContext` data the offering being purchased as well. I checked the implementation for Android and decided to leave it as a followup, and I just have offering identifier and displayCloseButton for now